File: src/DataBase_Reader_VB/Table Edits/MultiCellEdit.vb
Public Class MultiCellEdit
    Inherits TableEdit

    Private ReadOnly _cellEdits() As CellEdit
    Private _minColumnIndex As Int32
    Private _maxColumnIndex As Int32
    Private _minRowIndex As Int32
    Private _maxRowIndex As Int32
    Public ReadOnly Property CellEdits() As CellEdit()
        Get
            Return _cellEdits
        End Get
    End Property
    Public Sub New(cellEdits() As CellEdit)
        _cellEdits = cellEdits
        _minColumnIndex = cellEdits(0).ColumnIndex
        _maxColumnIndex = _minColumnIndex
        _minRowIndex = cellEdits(0).RowIndex
        _maxRowIndex = _minRowIndex
        '
        For Each edit As CellEdit In cellEdits
            If _minColumnIndex > edit.ColumnIndex Then _minColumnIndex = edit.ColumnIndex
            If _maxColumnIndex < edit.ColumnIndex Then _maxColumnIndex = edit.ColumnIndex
            If _minRowIndex > edit.RowIndex Then _minRowIndex = edit.RowIndex
            If _maxRowIndex < edit.RowIndex Then _maxRowIndex = edit.RowIndex
        Next
    End Sub

    Public Overrides Function GetEditedCellsInColumn(indexOfColumn As Integer) As List(Of CellEdit)
        Dim result As New List(Of CellEdit)
        If indexOfColumn < _minColumnIndex Then Return result
        If indexOfColumn > _maxColumnIndex Then Return result
        For i As Int32 = 0 To _cellEdits.Count - 1
            If _cellEdits(i).ColumnIndex = indexOfColumn And _cellEdits(i).RowIndex >= 0 Then result.Add(_cellEdits(i))
        Next
        Return result
    End Function

    Public Overrides Function GetEditedCellsInRow(indexOfRow As Integer) As List(Of CellEdit)
        Dim result As New List(Of CellEdit)
        If indexOfRow < _minRowIndex Then Return result
        If indexOfRow > _maxRowIndex Then Return result
        For i As Int32 = 0 To _cellEdits.Count - 1
            If _cellEdits(i).RowIndex = indexOfRow And _cellEdits(i).ColumnIndex >= 0 Then result.Add(_cellEdits(i))
        Next
        Return result
    End Function

    Public Overrides Sub ColumnAdded(Of T)(indexOfColumn As Integer, Optional ByVal columnData() As T = Nothing)
        If indexOfColumn > _maxColumnIndex Then Exit Sub
        For i As Int32 = 0 To _cellEdits.Count - 1
            _cellEdits(i).ColumnAdded(indexOfColumn, columnData)
        Next
        If indexOfColumn <= _minColumnIndex Then _minColumnIndex += 1
        If indexOfColumn <= _maxColumnIndex Then _maxColumnIndex += 1
    End Sub

    Public Overrides Sub ColumnDeleted(indexOfColumn As Integer)
        If indexOfColumn > _maxColumnIndex Then Exit Sub
        For i As Int32 = 0 To _cellEdits.Count - 1
            _cellEdits(i).ColumnDeleted(indexOfColumn)
        Next
        If indexOfColumn < _minColumnIndex Then _minColumnIndex -= 1
        If indexOfColumn < _maxColumnIndex Then _maxColumnIndex -= 1
    End Sub

    Public Overrides Sub RowAdded(indexOfRow As Integer, Optional ByVal rowData() As Object = Nothing)
        If indexOfRow > _maxRowIndex Then Exit Sub
        For i As Int32 = 0 To _cellEdits.Count - 1
            _cellEdits(i).RowAdded(indexOfRow)
        Next
        If indexOfRow <= _minRowIndex Then _minRowIndex += 1
        If indexOfRow <= _maxRowIndex Then _maxRowIndex += 1
    End Sub

    Public Overrides Sub RowDeleted(indexOfRow As Integer)
        If indexOfRow > _maxRowIndex Then Exit Sub
        For i As Int32 = 0 To _cellEdits.Count - 1
            _cellEdits(i).RowDeleted(indexOfRow)
        Next
        If indexOfRow < _minRowIndex Then _minRowIndex -= 1
        If indexOfRow < _maxRowIndex Then _maxRowIndex -= 1
    End Sub
    Public Overrides Function ContainsCell(indexOfColumn As Integer, indexOfRow As Integer, ByRef returnValue As Object) As Boolean
        If indexOfRow < _minRowIndex Then Return False
        If indexOfRow > _maxRowIndex Then Return False
        If indexOfColumn < _minColumnIndex Then Return False
        If indexOfColumn > _maxColumnIndex Then Return False
        For i As Int32 = 0 To _cellEdits.Count - 1
            If _cellEdits(i).ContainsCell(indexOfColumn, indexOfRow, returnValue) Then Return True
        Next
        Return False
    End Function
    Public Overrides Function ContainsColumn(indexOfColumn As Integer) As Boolean
        If indexOfColumn < _minColumnIndex Then Return False
        If indexOfColumn > _maxColumnIndex Then Return False
        For i As Int32 = 0 To _cellEdits.Count - 1
            If _cellEdits(i).ColumnIndex = indexOfColumn Then Return True
        Next
        Return False
    End Function
    Public Overrides Function ContainsRow(indexOfRow As Integer) As Boolean
        If indexOfRow < _minRowIndex Then Return False
        If indexOfRow > _maxRowIndex Then Return False
        For i As Int32 = 0 To _cellEdits.Count - 1
            If _cellEdits(i).RowIndex = indexOfRow Then Return True
        Next
        Return False
    End Function
End Class

File: src/DatabaseManager/Table Edits/MultiCellEdit.cs
/*
* NOTICE:
* The U.S. Army Corps of Engineers, Risk Management Center (USACE-RMC) makes no guarantees about
* the results, or appropriateness of outputs, obtained from this software.
*
* LIST OF CONDITIONS:
* Redistribution and use in source and binary forms, with or without modification, are permitted
* provided that the following conditions are met:
* ● Redistributions of source code must retain the above notice, this list of conditions, and the
* following disclaimer.
* ● Redistributions in binary form must reproduce the above notice, this list of conditions, and
* the following disclaimer in the documentation and/or other materials provided with the distribution.
* ● The names of the U.S. Government, the U.S. Army Corps of Engineers, the Institute for Water
* Resources, or the Risk Management Center may not be used to endorse or promote products derived
* from this software without specific prior written permission. Nor may the names of its contributors
* be used to endorse or promote products derived from this software without specific prior
* written permission.
*
* DISCLAIMER:
* THIS SOFTWARE IS PROVIDED BY THE U.S. ARMY CORPS OF ENGINEERS RISK MANAGEMENT CENTER
* (USACE-RMC) "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL USACE-RMC BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System;
using System.Collections.Generic;

namespace DatabaseManager
{
    /// <summary>
    /// Represents an edit that modifies multiple individual cells within a table.
    /// Provides optimized range-based access for row and column edits.
    /// </summary>
    public class MultiCellEdit : TableEdit
    {

        private readonly CellEdit[] _cellEdits;
        private int _minColumnIndex;
        private int _maxColumnIndex;
        private int _minRowIndex;
        private int _maxRowIndex;

        /// <summary>
        /// Gets the array of individual cell edits contained in this group edit.
        /// </summary>
        public CellEdit[] CellEdits
        {
            get
            {
                return _cellEdits;
            }
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="MultiCellEdit"/> class.
        /// </summary>
        /// <param name="cellEdits">An array of <see cref="CellEdit"/> objects representing the edited cells.</param>
        public MultiCellEdit(CellEdit[] cellEdits)
        {
            if (cellEdits == null || cellEdits.Length == 0) throw new ArgumentException("cellEdits cannot be null or empty.", nameof(cellEdits));
            _cellEdits = cellEdits;
            _minColumnIndex = cellEdits[0].ColumnIndex;
            _maxColumnIndex = _minColumnIndex;
            _minRowIndex = cellEdits[0].RowIndex;
            _maxRowIndex = _minRowIndex;
            // 
            foreach (CellEdit edit in cellEdits)
            {
                if (_minColumnIndex > edit.ColumnIndex)
                    _minColumnIndex = edit.ColumnIndex;
                if (_maxColumnIndex < edit.ColumnIndex)
                    _maxColumnIndex = edit.ColumnIndex;
                if (_minRowIndex > edit.RowIndex)
                    _minRowIndex = edit.RowIndex;
                if (_maxRowIndex < edit.RowIndex)
                    _maxRowIndex = edit.RowIndex;
            }
        }

        /// <summary>
        /// Returns the list of cell edits that occur in the specified column.
        /// </summary>
        /// <param name="indexOfColumn">The column index to search.</param>
        /// <returns>A list of <see cref="CellEdit"/> objects in the specified column.</returns>
        public override List<CellEdit> GetEditedCellsInColumn(int indexOfColumn)
        {
            var result = new List<CellEdit>();
            if (indexOfColumn < _minColumnIndex)
                return result;
            if (indexOfColumn > _maxColumnIndex)
                return result;
            for (int i = 0; i < _cellEdits.Length; i++)
            {
                if (_cellEdits[i].ColumnIndex == indexOfColumn && _cellEdits[i].RowIndex >= 0)
                    result.Add(_cellEdits[i]);
            }
            return result;
        }

        /// <summary>
        /// Returns the list of cell edits that occur in the specified row.
        /// </summary>
        /// <param name="indexOfRow">The row index to search.</param>
        /// <returns>A list of <see cref="CellEdit"/> objects in the specified row.</returns>
        public override List<CellEdit> GetEditedCellsInRow(int indexOfRow)
        {
            var result = new List<CellEdit>();
            if (indexOfRow < _minRowIndex)
                return result;
            if (indexOfRow > _maxRowIndex)
                return result;
            for (int i = 0; i < _cellEdits.Length; i++)
            {
                if (_cellEdits[i].RowIndex == indexOfRow && _cellEdits[i].ColumnIndex >= 0)
                    result.Add(_cellEdits[i]);
            }
            return result;
        }

        /// <summary>
        /// Updates all contained <see cref="CellEdit"/> instances in response to a column addition .
        /// </summary>
        /// <typeparam name="T">The type of the added column data.</typeparam>
        /// <param name="indexOfColumn">The index where the column was added.</param>
        /// <param name="columnData">Optional column data.</param>
        public override void ColumnAdded<T>(int indexOfColumn, T[] columnData = null)
        {
            if (indexOfColumn > _maxColumnIndex)
                return;
            for (int i = 0; i < _cellEdits.Length; i++)
                _cellEdits[i].ColumnAdded(indexOfColumn, columnData);
            if (indexOfColumn <= _minColumnIndex)
                _minColumnIndex += 1;
            if (indexOfColumn <= _maxColumnIndex)
                _maxColumnIndex += 1;
        }

        /// <summary>
        /// Updates all contained<see cref="CellEdit"/> instances in response to a column deletion
        /// </summary>
        /// <param name="indexOfColumn">The index of the column that was deleted.</param>
        public override void ColumnDeleted(int indexOfColumn)
        {
            if (indexOfColumn > _maxColumnIndex)
                return;
            for (int i = 0; i < _cellEdits.Length; i++)
                _cellEdits[i].ColumnDeleted(indexOfColumn);
            if (indexOfColumn < _minColumnIndex)
                _minColumnIndex -= 1;
            if (indexOfColumn < _maxColumnIndex)
                _maxColumnIndex -= 1;
        }

        /// <summary>
        /// Updates all contained <see cref="CellEdit"/> instances in response to a row addition.
        /// </summary>
        /// <param name="indexOfRow">The index where the row was added.</param>
        /// <param name="rowData">Optional row data.</param>
        public override void RowAdded(int indexOfRow, object[] rowData = null)
        {
            if (indexOfRow > _maxRowIndex)
                return;
            for (int i = 0; i < _cellEdits.Length; i++)
                _cellEdits[i].RowAdded(indexOfRow);
            if (indexOfRow <= _minRowIndex)
                _minRowIndex += 1;
            if (indexOfRow <= _maxRowIndex)
                _maxRowIndex += 1;
        }

        /// <summary>
        /// Updates all contained <see cref="CellEdit"/> instances in response to a row deletion.
        /// </summary>
        /// <param name="indexOfRow">The index of the row that was deleted.</param>
        public override void RowDeleted(int indexOfRow)
        {
            if (indexOfRow > _maxRowIndex)
                return;
            for (int i = 0; i < _cellEdits.Length; i++)
                _cellEdits[i].RowDeleted(indexOfRow);
            if (indexOfRow < _minRowIndex)
                _minRowIndex -= 1;
            if (indexOfRow < _maxRowIndex)
                _maxRowIndex -= 1;
        }

        /// <summary>
        /// Checks whether the specified cell is included in the edit and retrieves its value if found.
        /// </summary>
        /// <param name="indexOfColumn">The column index of the cell.</param>
        /// <param name="indexOfRow">The row index of the cell.</param>
        /// <param name="returnValue">The value of the cell if found.</param>
        /// <returns><c>true</c> if the specified cell is included in this edit; otherwise, <c>false</c>.</returns>
        public override bool ContainsCell(int indexOfColumn, int indexOfRow, ref object returnValue)
        {
            if (indexOfRow < _minRowIndex)
                return false;
            if (indexOfRow > _maxRowIndex)
                return false;
            if (indexOfColumn < _minColumnIndex)
                return false;
            if (indexOfColumn > _maxColumnIndex)
                return false;
            for (int i = 0; i < _cellEdits.Length; i++)
            {
                if (_cellEdits[i].ContainsCell(indexOfColumn, indexOfRow, ref returnValue))
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Checks whether the edit contains any changes in the specified column.
        /// </summary>
        /// <param name="indexOfColumn">The column index to check.</param>
        /// <returns><c>true</c> if the edit contains changes in the specified column; otherwise, <c>false</c>.</returns>
        public override bool ContainsColumn(int indexOfColumn)
        {
            if (indexOfColumn < _minColumnIndex)
                return false;
            if (indexOfColumn > _maxColumnIndex)
                return false;
            for (int i = 0; i < _cellEdits.Length; i++)
            {
                if (_cellEdits[i].ColumnIndex == indexOfColumn)
                    return true;
            }
            return false;
        }

        /// <summary>
        /// Checks whether the edit contains any changes in the specified row.
        /// </summary>
        /// <param name="indexOfRow">The row index to check.</param>
        /// <returns><c>true</c> if the edit contains changes in the specified row; otherwise, <c>false</c>.</returns>
        public override bool ContainsRow(int indexOfRow)
        {
            if (indexOfRow < _minRowIndex)
                return false;
            if (indexOfRow > _maxRowIndex)
                return false;
            for (int i = 0; i < _cellEdits.Length; i++)
            {
                if (_cellEdits[i].RowIndex == indexOfRow)
                    return true;
            }
            return false;
        }
    }
}